### DISCUSSION TITLE
qudi scannig and poi module with attocube piezo stepper stage

### DISCUSSION THREAD

**Discussion by Janoo12** [Ulm-IQO/qudi-core#97] [Ideas]
I was wondering whether the scanning probe toolchain could be extended in a way that it can be used with Attocube piezo stepper stage (ECSxy5050/Al/NUM/RT + ECSz5050 + AMC100 controller) in closed loop mode for image making.  The ambient temperature stepper stages have highly precise (1-10 nm) encoders, thus in closed loop function it's possible to go point by point and recording counts or carrying out other types of measurements at any point (like ODMR or lifetime measurements) and creating an image from that information. The main difference that the control of a stepper stage doesn't happen via analog outputs but via USB commands, but Attocube provides python commands to the controller so it could be implemented. 
This extension could be particularly useful for those setups:
- which don't have an analog scanner but still want imaging even if image making this way takes more time
- extended measurements could be carried out point by point without any movement which could possibly disturb those measurements

**Comment by LukePiewalker** [comment]
Dear @Janoo12,

from the toolchain perspective, it is kept so general that this could be implemented.
It is on such a high level, that all the necessary things would need to be taken care on the "hardware level".
However, one would need to program the hardware accordingly to take care of the timing of the everything: stepping, creating the data values of the image. Maybe one could make that as an "interfuse"?

There has been an earlier question, using attocubes for the z-scan [here](https://github.com/Ulm-IQO/qudi-core/discussions/53). So I guess some implementation is of interest, yet we still need a brave soul to come up with an implementation and maybe a clever idea to make such a hardware pretty general.

**Comment by AndrePointner** [comment]
Hi,

we also use a attocube stage. Unfortunately the confocal scanning logic is so tightly fused with the NIDAQ card, that we had to make quite a lot of modifications to it. But it worked out - once we broke down the workings of the interfuse, we knew where to insert the Python Commands for the Stage Control. We basically replaced the Analog Control via the NIDAQ with the according Python Commands for the Stage.

At the end of the day it is only a few lines of modification to get it working - maybe not according to the Best-Practice-Standard for QUDI, but works.

BR

**Reply by Janoo12** [reply]
That sounds very good. Is it possible to share those modified scripts with the community in some way?

**Comment by Neverhorst** [comment]
Hi @Hundsraudigerrotzbua,

Just to avoid possible misunderstandings:
Are you using the latest version of the `qudi-iqo-modules` package for your scanning implementation?
I'm just asking because you mention "confocal scanning logic" and that it is tightly coupled to the NI hardware. 
We re-developed the entire scanning toolchain specifically to get rid of this issue and named the new scanner controlling logic [scanning_probe_logic.py](https://github.com/Ulm-IQO/qudi-iqo-modules/blob/main/src/qudi/logic/scanning_probe_logic.py).

If you are in fact using this new toolchain, could you please help us understand where the logic is still coupled to NI hardware? 

If you are referring to [ni_scanning_probe_interfuse.py](https://github.com/Ulm-IQO/qudi-iqo-modules/blob/main/src/qudi/hardware/interfuse/ni_scanning_probe_interfuse.py), this is a specialized "hardware implementation" for NI hardware and as such it is not meant to be generally applicable. Different scanning setups are meant to implement their own hardware/interfuse module satisfying the very general interface in [scanning_probe_interface.py](https://github.com/Ulm-IQO/qudi-iqo-modules/blob/main/src/qudi/interface/scanning_probe_interface.py) instead.

Thank you.

**Reply by AndrePointner** [reply]
You're right, we modified the interfuse. I was mistaken, as I worked a lot on the older confocal scanning logic. With qudi-core we modified the iqo-module that combines the two NIDAQ hardware files to also include the functionality for the piezo. 

Unfortunately there is not much to upload from our part, as the colleague of mine working on with the piezo didn't follow up on the implementation and it is not yet functional.

All I can say is, that I remember modifying the ```__ao_cursor_write_loop()``` in the ```ni_scanning_probe_interfuse``` to also move the stack respectively.

Thanks for clearing that up!
